p2panda-rs/src/schema/schema_id.rs

```diff
@@ -32,20 +33,71 @@ impl SchemaId {
         match id {
             "schema_v1" => Ok(SchemaId::Schema),
             "schema_field_v1" => Ok(SchemaId::SchemaField),
-            hash_str => Ok(hash_str.parse::<DocumentViewId>()?.into()),
+            application_schema_id => Self::parse_application_schema_str(application_schema_id),
         }
     }
-}
 
-impl From<OperationId> for SchemaId {
-    fn from(operation_id: OperationId) -> Self {
-        Self::Application(PinnedRelation::new(operation_id.into()))
+    /// Returns a `SchemaId` given an application schema's name and view id.
+    pub fn new_application(name: &str, view_id: &DocumentViewId) -> Self {
+        SchemaId::Application(name.to_string(), view_id.clone())
+    }
+
+    /// Read an application schema id from a string.
+    ///
+    /// Parses the schema id by iteratively splitting sections from the right at `_` until the
+    /// remainder is shorter than an operation id. Each section is parsed as an operation id
+    /// and the last (leftmost) section is parsed as the schema's name.
+    fn parse_application_schema_str(id_str: &str) -> Result<Self, SchemaIdError> {
+        if id_str.find('_').is_none() {
+            return Err(SchemaIdError::MalformedApplicationSchemaId(
+                "expecting name and view id hashes separated by underscore".to_string(),
+            ));
+        }
+
+        let mut operation_ids = vec![];
+        let mut remainder = id_str;
+        while let Some((left, right)) = remainder.rsplit_once('_') {
+            // Catch trying to parse an unknown system schema
+            if right.starts_with('v') && right.len() < MAX_YAMF_HASH_SIZE * 2 {
+                return Err(SchemaIdError::UnknownSystemSchema(id_str.to_string()));
+            }
```

Comment:
So this is catching the case where a system schema which wasn't matched against an existing system schema id and was passed then passed into `parse_application_schema_str`? Coming at this cold, the "unknown system schema" confused me a little, I would export more something along the lines of "invalid application schema". Could the "unknown system schema" error be caught elsewhere and inside this method we only deal with errors related to application schema id parsing?

Reply:
This could completely be left out, then you'd just get an error about it not being able to parse an operation id. I thought it'd be nice to catch that situation here and present a more helpful error message. If we wanted to catch it outside the method we would have to have an additional parser in the constructor (on the application schema match arm) just to check for that. I thought that wouldn't be worth it.

Reply:
How about I add a comment that explains this situation?

Reply:
Maybe it could be split into a separate method and then used as a condition in the match arms?

Reply:
Not sure what you mean :D Want to hop on Discord?

Reply:
Yeh, maybe this is the best way. Obvs nicer error messages is optimal, I only got confused understanding the code, and a comment would clear that up for sure.